Be resilient to options being serialized as strings.

CyrusNajmabadi opened the pull request:
**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:**

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)

CyrusNajmabadi commented:
Tagging @dotnet/roslyn-ide @pilchie

CyrusNajmabadi commented:
Tagging @dotnet/roslyn-ide @pilchie

sharwell commented on src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs:
```diff
@@ -96,6 +95,27 @@ bool IOptionPersister.TryFetch(OptionKey optionKey, out object value)
+
+                        if (untypedValue is string stringValue)
+                        {
+                            // Due to a previous bug we were accidentally serializing longs as strings. Gracefully convert
```
❓ How does this apply to `int` values?

sharwell commented on src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs:
```diff
@@ -77,24 +77,41 @@ bool IOptionPersister.TryFetch(OptionKey optionKey, out object value)
+                    }
+                    else if (optionKey.Option.Type == typeof(int))
+                    {
+                        var untypedValue = subKey.GetValue(key, defaultValue: optionKey.Option.DefaultValue);
```
❔ Why not use `Convert.ToInt32(untypedValue)`?

CyrusNajmabadi replied:
I think we want to be super crisp about how things convert to/from registry form.  And, in the case where we've stored data in the wrong form, we want to be crisp about the code that gracefully handles it.

CyrusNajmabadi commented:
@jasonmalinowski Despite your fix, i still got this (presumably due to some bad cached values locally on my box.  So i expanded your fix to not just handle 'longs' but also 'ints' as well.

CyrusNajmabadi commented:
Tagging @MattGertz 

**Customer scenario**

The recent change to add hte goldbar to offer the new 'fxcop vsix experience' ended up serializing some data badly into the registry.  This change makes it so we gracefully handle that case for machines that may have run that code in the interim period before we fixed that issue.

**Bugs this fixes:**

None.

**Workarounds, if any**

Delete specific registry keys.  No way to know easily know which you'd have to delete.  Until you fix it VS just crashes.

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

The original code to store 'gold bar' data was tested locally without actually shutting down VS.  So we never exercised the codepaths that actually persisted to the registry and back.  Instead, we only really tested in memory representation.  The registry codepaths were not correct, and you could get a failure to deserialize the data.  We've fixed this up, but machines that participated in this small window of time may still have bad data serialized.  This makes us resilient ot that so that these users will not crash.

**How was the bug found?**

Local dogfooding.
